File: study/ts/script/scene/view/homeland/HomelandFriendView.ts
const { ccclass, property } = cc._decorator;

import CommonTopbarBase from '../../../ui/component/CommonTopbarBase'

import CommonHeroPower from '../../../ui/component/CommonHeroPower'

import CommonMainMenu from '../../../ui/component/CommonMainMenu'
import ViewBase from '../../ViewBase';
import { TopBarStyleConst } from '../../../const/TopBarStyleConst';
import { FunctionConst } from '../../../const/FunctionConst';
import { handler } from '../../../utils/handler';
import { G_UserData, Colors, G_SceneManager } from '../../../init';
import { HomelandHelp } from './HomelandHelp';
import { HomelandConst } from '../../../const/HomelandConst';
import HomelandNode from './HomelandNode';
import HomelandMainNode from './HomelandMainNode';
import HomelandGuildList from './HomelandGuildList';
import UIHelper from '../../../utils/UIHelper';
import HomelandPrayNode from './HomelandPrayNode';

@ccclass
export default class HomelandFriendView extends ViewBase {

    @property({
        type: cc.Node,
        visible: true
    })
    _panelDesign: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _mainNode: cc.Node = null;

    @property({
        type: CommonMainMenu,
        visible: true
    })
    _commonBack: CommonMainMenu = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _playerName: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _playerTxt: cc.Label = null;

    @property({
        type: CommonHeroPower,
        visible: true
    })
    _fileNodePower: CommonHeroPower = null;

    @property({
        type: cc.Button,
        visible: true
    })
    _buttonFold: cc.Button = null;

    @property({
        type: CommonTopbarBase,
        visible: true
    })
    _topbarBase: CommonTopbarBase = null;
    @property({
        type: cc.Prefab,
        visible: true
    })
    homelandNode: cc.Prefab = null;
    @property({
        type: cc.Prefab,
        visible: true
    })
    homelandMainNode: cc.Prefab = null;
    @property({
        type: cc.Prefab,
        visible: true
    })
    homelandGuildList: cc.Prefab = null;
    @property({
        type: cc.Prefab,
        visible: true
    })
    homelandPrayNode: cc.Prefab = null;

    _friendId: any;
    _guildListView: any;
    _mainTree: HomelandMainNode;
    _guildListSignal: any;
    _popResult: any;

    onCreate() {
        var params = G_SceneManager.getViewArgs('homelandFriend');
        this._friendId = params[0];
        this._guildListView = null;

        this.setSceneSize();
        this.updateSceneId(2001);
        this._makeBackGroundBottom();
        this._updateTreeModel();
        this._initPrayModel();
        this._commonBack.updateUI(FunctionConst.FUNC_HOMELAND_BACK);
        this._commonBack.addClickEventListenerEx(handler(this, this.onBackToMain));
        this._updatFriendName();
    }
    start() {
        this._topbarBase.updateUI(TopBarStyleConst.STYLE_HOMELAND);
        this._topbarBase.setImageTitle('txt_big_homeland_01');
    }
    _updatFriendName() {
        var guildData = G_UserData.getHomeland().getGuildMemberByFriendId(this._friendId);
        if (guildData) {
            var official = guildData.getOfficer_level();
            this._playerName.node.color = (Colors.getOfficialColor(official));
            UIHelper.enableOutline(this._playerName, Colors.getOfficialColorOutline(official), 1);
            this._playerName.string = (guildData.getName());
            this._playerName['_updateRenderData'](true);
            var widthName = this._playerName.node.getContentSize().width;
            var widthTxt = this._playerTxt.node.getContentSize().width;
            var center = (widthName + widthTxt) / 2;
            this._playerName.node.x = (widthName - center);
            this._playerTxt.node.x = (widthName - center + 8);
        }
    }
    _updatePower() {
        this._fileNodePower.updateUI(HomelandHelp.getFriendAllPower(this._friendId));
        this._fileNodePower.node.x = -this._fileNodePower.getWidth() * 0.5;
    }
    _updateTreeModel() {
        var groundNode = this.getGroundNode();
        for (var i = 1; i <= HomelandConst.MAX_SUB_TREE; i++) {
            var subTree = G_UserData.getHomeland().getInviteFriendSubTree(this._friendId, i);
            var subNode = groundNode.getChildByName('subNode' + i);
            if (subNode == null) {
                subNode = cc.instantiate(this.homelandNode);
                let comp: HomelandNode = subNode.getComponent(HomelandNode);
                comp.ctor(HomelandConst.FRIEND_TREE)
                subNode.name = ('subNode' + i);
                groundNode.addChild(subNode);
                this['_subTree' + i] = comp;
            }
            if (subTree && subNode) {
                subNode.getComponent(HomelandNode).updateUI(subTree);
            }
        }
        var mainTree = G_UserData.getHomeland().getInviteFriendMainTree(this._friendId);
        var mainNode = groundNode.getChildByName('mainTree');
        if (mainNode == null && mainTree) {

            mainNode = cc.instantiate(this.homelandMainNode);
            this._mainTree = mainNode.getComponent(HomelandMainNode)
            this._mainTree.ctor(HomelandConst.FRIEND_TREE)
            groundNode.addChild(mainNode);
            mainNode.name = ('mainTree');
        }
        this._mainTree.updateUI(mainTree);
    }

    _initPrayModel () {
        var groundNode = this.getGroundNode();
        var prayNode = cc.instantiate(this.homelandPrayNode).getComponent(HomelandPrayNode);
        prayNode.ctor(HomelandConst.FRIEND_TREE, null);
        prayNode.updateRedPoint();
        groundNode.addChild(prayNode.node);
    }

    onButtonFold() {
        this._createGuildList();
    }
    _createGuildList() {
        if (this._guildListView == null) {
            var popup = cc.instantiate(this.homelandGuildList).getComponent(HomelandGuildList);
            popup.ctor(this._friendId);
            popup.node.name = ('HomelandGuildList');
            this.node.addChild(popup.node);
            this._guildListView = popup;
            this._guildListSignal = this._guildListView.signal.add(handler(this, this._onGuildListClose));
            this._popResult = null;
            this._buttonFold.node.active = (false);
        }
    }
    onEnter() {
        this._updatePower();
        this._createGuildList();
    }
    onExit() {
    }
    onBackToMain() {
        G_SceneManager.replaceScene('homeland');
    }
    _onGuildListClose(event) {
        if (event == 'close') {
            this._buttonFold.node.active = (true);
            this._guildListView = null;
            if (this._guildListSignal) {
                this._guildListSignal.remove();
                this._guildListSignal = null;
            }
        }
    }
}

File: study/lua/src/app/scene/view/homelandFriend/HomelandFriendView.lua

-- Author: hedili
-- Date:2018-04-19 14:10:14
-- Describle：

local ViewBase = require("app.ui.ViewBase")
local HomelandFriendView = class("HomelandFriendView", ViewBase)
local HomelandMainNode = require("app.scene.view.homeland.HomelandMainNode")
local HomelandNode = require("app.scene.view.homeland.HomelandNode")
local HomelandGuildList = require("app.scene.view.homeland.HomelandGuildList")
local HomelandHelp = require("app.scene.view.homeland.HomelandHelp")
local HomelandConst = require("app.const.HomelandConst")
local HomelandPrayNode = require("app.scene.view.homeland.HomelandPrayNode")

function HomelandFriendView:ctor(friendId)
	dump(friendId)
	--csb bind var name
	self._commonChat = nil  --CommonMiniChat
	self._commonHelp = nil  --CommonHelp
	self._panelbk = nil  --Panel
	self._topbarBase = nil  --CommonTopbarBase
	self._betIcon = nil 	--投注按钮
	self._mainNode = nil 
	self._buttonFold = nil --
	self._friendId = friendId
	self._guildListView = nil
	local resource = {
		file = Path.getCSB("HomelandFriendView", "homeland"),
		size = G_ResolutionManager:getDesignSize(),
		binding = {
			_buttonFold = {events = {{event = "touch", method = "_onButtonFold"}}},
		},
	}
	HomelandFriendView.super.ctor(self, resource, 2001)
end

-- Describle：

function HomelandFriendView:onCreate()
	self:_makeBackGroundBottom()
	self:_updateTreeModel()
	self:_initPrayModel()
	local TopBarStyleConst = require("app.const.TopBarStyleConst")
	self._topbarBase:updateUI(TopBarStyleConst.STYLE_HOMELAND)
	self._topbarBase:setImageTitle("txt_big_homeland_01")

	self._commonBack:updateUI(FunctionConst.FUNC_HOMELAND_BACK)
	self._commonBack:addClickEventListenerEx(handler(self, self.onBackToMain))
	
	self:_updatFriendName()
end


function HomelandFriendView:_updatFriendName( ... )
	-- body
	local guildData = G_UserData:getHomeland():getGuildMemberByFriendId(self._friendId)
	if guildData then
		local official = guildData:getOfficer_level()
		self._playerName:setColor(Colors.getOfficialColor(official))
		self._playerName:enableOutline(Colors.getOfficialColorOutline(official),1)
		self._playerName:setString(guildData:getName())
		local widthName = self._playerName:getContentSize().width
		local widthTxt = self._playerTxt:getContentSize().width
		local center = (widthName + widthTxt) / 2
		self._playerName:setPositionX(widthName - center)
		self._playerTxt:setPositionX(widthName - center + 8)
	end
	
end

function HomelandFriendView:_updatePower( ... )
	-- body
	self._fileNodePower:updateUI(HomelandHelp.getFriendAllPower(self._friendId))
	self._fileNodePower:setPositionX(250 - self._fileNodePower:getWidth()*0.5)
	
end

function HomelandFriendView:_updateTreeModel( ... )

	local groundNode = self:getGroundNode()

	for i= 1, HomelandConst.MAX_SUB_TREE do
		local subTree = G_UserData:getHomeland():getInviteFriendSubTree(self._friendId, i)

		local subNode = groundNode:getChildByName("subNode"..i)
		if subNode == nil  then
			subNode = HomelandNode.new(HomelandConst.FRIEND_TREE)
			subNode:setName("subNode"..i)
			groundNode:addChild(subNode)
			self["_subTree"..i] = subNode
		end
		if subTree and subNode then
			dump(subTree)
			subNode:updateUI(subTree)
		end
	end

	-- body

	local mainTree = G_UserData:getHomeland():getInviteFriendMainTree(self._friendId)
	local mainNode = groundNode:getChildByName("mainTree")
	if mainNode == nil and mainTree then
		mainNode = HomelandMainNode.new(HomelandConst.FRIEND_TREE)
		groundNode:addChild(mainNode)
		mainNode:setName("mainTree")
	end
	mainNode:updateUI(mainTree)

	--mainNode:setPosition(HomelandConst.MAIN_TREE_POSITION)
	self._mainTree = mainNode
end

function HomelandFriendView:_initPrayModel()
	local groundNode = self:getGroundNode()
	local prayNode = HomelandPrayNode.new(HomelandConst.FRIEND_TREE)
	groundNode:addChild(prayNode)
	prayNode:updateRedPoint()
end

-- Describle：
function HomelandFriendView:_onButtonFold()
	-- body
	logWarn("HomelandFriendView:_onButtonFold")

	self:_createGuildList()

end

function HomelandFriendView:_createGuildList( ... )
	-- body
	if self._guildListView == nil then
		local popup = HomelandGuildList.new(self._friendId)
		popup:setName("HomelandGuildList")
		self:addChild(popup)
		self._guildListView = popup
		self._guildListSignal = self._guildListView.signal:add(handler(self, self._onGuildListClose))
		self._popResult = nil
		self._buttonFold:setVisible(false)
	end
end
-- Describle：
function HomelandFriendView:onEnter()
	logWarn("HomelandFriendView")
	self:_updatePower()

	self:_createGuildList()
end

-- Describle：
function HomelandFriendView:onExit()

end

function HomelandFriendView:onBackToMain( ... )
	local scene, view = G_SceneManager:createScene("homeland")
	G_SceneManager:replaceScene(scene)
end


--公会列表关闭事件通知
function HomelandFriendView:_onGuildListClose(event)
    if event == "close" then
		self._buttonFold:setVisible(true)
        self._guildListView = nil
		if self._guildListSignal then
			self._guildListSignal:remove()
			self._guildListSignal = nil
		end
    end
end
return HomelandFriendView